### PR TITLE
Add notebooks for imaging mode

### PR DESCRIPTION
**Pull request by bhilbert4**
This PR replaces #2 as the PR for adding the imaging mode pipeline notebooks. #2 was pointing to the wrong branch, and my local clone was created long enough ago that the `main` (and even `master`) branch did not exist. So I re-cloned the repo and am re-creating the PR. The notebooks are identical to how they were in #2. More changes have yet to be made based on feedback from the Data Products webinar dry run.

### CODE REVIEW

**Review by jotaylor — COMMENTED** [review]
In the stage1 notebook
Introduction section:

-  "The output is a corrected but uncalibrated countrate or slope image". this statement is a little misleading. Or maybe I don't know what the word "calibrated" actually means...
- can we include a link to the up-the-ramp jdox article in the intro? https://jwst-docs.stsci.edu/understanding-exposure-times#UnderstandingExposureTimes-uptherampHowup-the-rampreadoutswork

**Comment by bhilbert4** [comment]
True. I should elaborate a bit on calibrated/uncalibrated. Following the terminology used in the calibration working group, and @karllark 's presentations, the files aren't "calibrated" until the end of the image2 pipeline, when the units change from adu/sec to MJy/str.

**Comment by jotaylor** [comment]
Sorry for the wall of text Bryan, it's mostly text tweaks. I was trying to go through this notebook from a new user's perspective so some of the comments may be basic.

**Comment by jotaylor** [comment]
**Stage1 notebook, I was looking at the static version**

Introduction:
- extra \ in `*\_rate.fits*` and `*\_cal.fits*`
- Different text format for _i2d.fits and _cal.fits in 3rd paragraph (italics vs code)

Pipeline Resources:
- same comment about unindenting Help Desk bullet

Installation: 
- text is slightly different from stage 1 NB. copy+paste one or the other perhaps?

Reference Files:
- text is slightly different from stage 1 NB. copy+paste one or the other perhaps?

Imports:
- same comment from Stage1 NB
- text in code cell below "import JWST pipeline-related modules" has references to calwebb_detector1 instead of calwebb_image2

Convenience functions:
- show image does not have pars in docstring

Download data
- no MIRI data yet

Association Files
- include text explaining why data should be grouped together and processed together. (take text from here? https://jwst-pipeline.readthedocs.io/en/stable/jwst/associations/overview.html#what-are-associations)

Methods for calling steps/pipelines
- same comments from Stage1 NB

The calwebb_image2 pipeline
- link to "Figure 1" doesn't work (https://jwst-docs.stsci.edu/jwst-data-reduction-pipeline/algorithm-documentation/stages-of-processing/calwebb_image2)
- same comments as calwebb_detector1 section in image1 NB

assign_wcs step:
- typo: "we'll loop over the input files, read them into **ImageCube** instances" -> "we'll loop over the input files, read them into **ImageModel** instances"
- suggest clarifying/reminding what output this is: "Let's show again the pipeline output image from above, and zoom in on an interesting area" -> "Let's look at the calwebb_image2 pipeline output image from before (the _cal.fits files), and zoom in on an interesting area"
- how did you get the X/Y locations of the sources? 

photom step:
- typo: "the **pixels** values in the science observation are multiplied by this factor" -> "the **pixel** values in the science observation are multiplied by this factor"
- typo: "also appended to the **observation** in a new extension" -> "also appended to the **observation file** in a new 
- extension"

resample step:
- suggest showing output from photom step compared to resample to emphasize the improvement

General comments:
- will users know about datamodels at this point? if not maybe a short paragraph about what they are, since datamodels are required for input in this NB
- steps are well explained and code illustrates calibration effects very well

**Comment by jotaylor** [comment]
**Stage2 notebook, I was looking at the static version**

Introduction:
- typo: "**Spurous** sources" -> "**Spurious** sources" 
- suggest reminding when jump was performed: "(e.g. cosmic rays that were not flagged in the jump step)" -> "(e.g. cosmic rays that were not flagged in the jump step during Stage 1 processing)"
- typo: "**udistorted** sky projection " -> "**undistorted** sky projection "

Pipeline resources, Installation, Reference files, Imports
- same comments as previous NBs
- text in code cell below "import JWST pipeline-related modules" has references to calwebb_detector1 instead of calwebb_image3

Convenience Functions:
- overlay_catalog and show_image do not have pars in their docstrings

Download data:
- no MIRI data

Association files:
- typo: "The same **asn** file that we created above " -> "The same **association** file that we created above "

Methods for calling steps/pipelines:
- same comments from Stage1 NB

Calwebb_image3 pipeline, using call() method:
- typo: "specify a **paramter** reference file" -> "specify a **parameter** reference file"
- there is no parameter reference file for image3, but the code cells try to use it. maybe these cells are meant to be deleted before the webinar?

tweakreg step:
- typo: " it **identfies** point sources" -> " it **identifies** point sources"
- suggest changing: "You'll see about halfway down the parameters field" -> "About halfway down, you'll see the parameters field"
- suggest changing: "The cell below is set to Raw. If you wish to run it and try the call() command, change the" -> "If you wish to run the call() command in the cell below, change the"
- typo: "WCS before and after step by loading" -> "WCS before and after the step by loading"

skymatch step:
- suggest rewording: "Typically these sky values are calculated such that it can find offsets that will minimize the difference in sky level between the input images when they are combined into a final mosaic." -> "Sky values can be computed for each image separately or in a way that matches the sky levels amongst the collection of images so as to minimize their differences." (taken verbatim from RTD)
- typo: "make a copy of an **exiting** association file " -> "make a copy of an **existing** association file "
- typo: " In **eacbh** file, the" -> " In **each** file, the"

outlier_detection step:
- something that confused me in my early JWST days was why bad pixels were flagged twice in the pipeline. It might be useful to touch on that briefly in the Summary section, i.e. why were some of these artifacts not flagged in the jump step, and why we can find them now.

source_catalog step:
- the last sentence in Summary got cut off: "The output"

**Comment by jotaylor** [comment]
@bhilbert4 I've finished my review of the notebooks.

**Comment by bhilbert4** [comment]
Thanks for the thorough review @jotaylor! I've made all the updates.

**Comment by bhilbert4** [comment]
@jdavies-st @hbushouse @nden Could someone give these a quick look-over and make sure that everything makes sense?

**Comment by bhilbert4** [comment]
Also, don't bother with the "live" versions of the notebooks. I'm going to wait to update those after the 3 other notebooks are finalized.

**Comment by jotaylor** [comment]
The new edits look great @bhilbert4 !!! I looked through and ran the notebooks one more time and have just a few more comments.

Stage1 NB:
- Run entire calwebb_detector1 section: the cell for the .call() method should be raw type , not code type.
- dark step: maybe make a similar note as calwebb_detector1 about how long this step can take?
- jump step: I forgot to note this last time, but one of the pixels marked as having jumps don't really seem to have jumps, when you look at the plots. You may get a question about this? (maybe that's intentional)
- ramp fitting: I'm still not getting slope=0 for the example pixel at idx=45. Printing `scipix[y, x]` gives 0.6169286. Even though the groupdq values are 2.

Stage2 NB:
- call the pipeline from the command line: has reference to calwebb_detector1

Steage3 NB:
- skymatch: typo, "make a copy of an **exsiting** association file" -> "make a copy of an **existing** association file"
- source_catalog: the last sentence in Summary is still cut off: "The output"

I really like the new formatting of the color MD cells.

**Comment by bhilbert4** [comment]
Thanks @jotaylor. I fixed those issues. For the jump step, I got tired of trying to cherry pick for really obvious jumps, and figured it might be better to show more options of what jumps may look like. I'm sure there will be questions about it, at which point I will direct them to the code and let them dive in. :)

**Review by jdavies-st — COMMENTED** [review]
Stage 1 notebook looks really good.  Just some very minor comments below.  Your choice on whether to incorporate them.

Will review stages 2 and 3 when I get time.

**Comment by nden** [comment]
@bhilbert4 These look very good! :tada:

**Comment by camipacifici** [comment]
Thank you all for the reviews! Need to merge this now to test the platform before the dry run tomorrow.
More comments can still come if needed.

**Comment by bhilbert4** [comment]
@camipacifici the live versions of the notebooks are not yet ready in this PR. Is there time for me to make copies of the static notebooks and delete some text in order to create the live versions (in a new PR)?

**Comment by camipacifici** [comment]
@bhilbert4 yes, please submit a new PR with the changes.
